### ISSUE TITLE
sensor logger shouldn't crash on bad insert

### ISSUE THREAD

**Issue by laroque** [project8/dragonfly#30]
we had a miss-named endpoint (middle_current_limit instead of middle_coil_current_limit I think) that produced the following traceback. Posting to slack is good (ie that a critical message posts) but the process should _not_ have crashed. It appears that there was a python IndexError which wasn't caught (I think because the query to the endpoint_id_map returned no results on the unknown sensor name).

```
2016-02-03T10:38:29[INFO    ] dripline.core.service(315) -> received a message
2016-02-03T10:38:29[WARNING ] dripline.core.gogol(51) -> got an exception:
list index out of range
2016-02-03T10:38:29[CRITICAL] dripline.core.service(414) -> Service <slow_control_table_interface> crashing with error message:
list index out of range
supposed to do an emit
2016-02-03T10:38:29[ERROR   ] dripline.core.service(415) -> Traceback (most recent call last):
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/dripline-wp2.0.0_python3.2.3_0_g1c92492-py2.7.egg/dripline/core/service.py", line 412, in run
    self._connection.ioloop.start()
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 138, in start
    self.poller.start()
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 433, in start
    self.poll()
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 494, in poll
    self._handler(fileno, event, write_only=write_only)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 322, in _handle_events
    self._handle_read()
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 351, in _handle_read
    self._on_data_available(data)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/connection.py", line 1285, in _on_data_available
    self._process_frame(frame_value)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/connection.py", line 1365, in _process_frame
    self._deliver_frame_to_channel(frame_value)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/connection.py", line 976, in _deliver_frame_to_channel
    return self._channels[value.channel_number]._handle_content_frame(value)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/channel.py", line 792, in _handle_content_frame
    self._on_deliver(*response)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/pika/channel.py", line 886, in _on_deliver
    body)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/dripline-wp2.0.0_python3.2.3_0_g1c92492-py2.7.egg/dripline/core/service.py", line 325, in on_message
    msg_type_handlers[message.msgtype](unused_channel, basic_deliver, properties, body)
  File "/home/project8/python_environments/drip3.2_dragon1.0/local/lib/python2.7/site-packages/dripline-wp2.0.0_python3.2.3_0_g1c92492-py2.7.egg/dripline/core/gogol.py", line 47, in on_alert_message
    self.this_consume(message_unpacked, method)
  File "build/bdist.linux-x86_64/egg/dragonfly/implementations/sensor_logger.py", line 53, in this_consume
    self._sensor_types[sensor_name] = this_type[1][0][0]
IndexError: list index out of range
```

**Comment by les67** [comment]
Database issues #27 #46 are related.

**Comment by les67** [comment]
@laroque in the case that the endpoint name is not found in the db, what is more appropriate: A logger.warning, logger.error, or? Unless there's a disagreement, I will set it up to give a logger error as I did for the SQLSnapshot methods. Something like <code>logger.error('endpoint with name "{}" not found in database'.format(name))</code>.

**Comment by guiguem** [comment]
Since it should be posted on slack, a logger.critical might be appropriate so it gets propagated to Slack with the AMQP handler and the Slack Interface.

**Comment by les67** [comment]
Oh ok that is a good idea then. I wasn't aware that all logger.critical messages get propagated to Slack. If this is so, then I will also change the logger.error messages I have in SQLSnapshot

**Comment by guiguem** [comment]
This is very recent (v1.3.0), I announced this this morning during the call.

**Comment by guiguem** [comment]
Be careful that the things that are sent to Slack should make some sense. An error which does not tell you what is going on or just tell some unintelligible sentence should be turned into something more explicit (like "the endpoint {} does not exist in the database; might need to add it". Also with this Slack interface when some service sends too many critical messages, Slack will ignore them for a given duration, so be careful that spamming on Slack will be kind of disabled.

**Comment by les67** [comment]
In bug/DatabaseIssues in SensorLogger, I am now catching the case when an endpoint name is not found in the db; I named it `misnamed endpoint`. This is the `if not this_type[1]:` line in the method this_consume. When this is the case I would like to send a critical message to slack. However, when I log a critical message I get the first error below. On the other hand, when I log an info message there is no issue; second error below. I would really like for this case to log a critical message so that it can be propagated to Slack. But, I'm not understanding why the error is present when logger.critical is used and not when logger.info is used, i.e. I don't understand why with the former case the `'Service' object has no attribute 'send_status_message'` @guiguem @laroque could you shed some light on this? 

1. A logger.critical is issued - **produces error**:

```
2016-11-16T20:53:29[INFO    ] dripline.core.service(316) -> received a message
2016-11-16T20:53:29[DEBUG   ] dripline.core.service(363) -> Acknowledging message 2
2016-11-16T20:53:29[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T20:53:29Z", "msgtype": 4, "payload": {"value_cal": 5.5, "value_raw": "2.75"}, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "54a27f7f29c6", "version": "v0.0.0", "commit": ""}}
2016-11-16T20:53:29[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T20:53:29Z', u'msgtype': 4, u'payload': {u'value_cal': 5.5, u'value_raw': u'2.75'}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'54a27f7f29c6', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T20:53:29[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T20:53:29[DEBUG   ] dripline.core.gogol(48) -> in process_message callback
2016-11-16T20:53:29[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T20:53:29Z", "msgtype": 4, "payload": {"value_cal": 5.5, "value_raw": "2.75"}, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "54a27f7f29c6", "version": "v0.0.0", "commit": ""}}
2016-11-16T20:53:29[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T20:53:29Z', u'msgtype': 4, u'payload': {u'value_cal': 5.5, u'value_raw': u'2.75'}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'54a27f7f29c6', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T20:53:29[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T20:53:29[CRITICAL] dragonfly.implementations.sensor_logger(59) -> endpoint with name "misnamed_endpoint" was not found in database hence failed to log its value; might need it to add to the db
2016-11-16T20:53:29[WARNING ] dripline.core.gogol(55) -> got an exception:
'Service' object has no attribute 'send_status_message'
2016-11-16T20:53:29[DEBUG   ] dripline.core.gogol(56) -> traceback follows:
Traceback (most recent call last):
  File "/code/dripline/python/dripline/core/gogol.py", line 51, in on_alert_message
    self.this_consume(message_unpacked, method)
  File "/code/dragonfly/dragonfly/implementations/sensor_logger.py", line 59, in this_consume
    logger.critical('endpoint with name "{}" was not found in database hence failed to log its value; might need it to add to the db'.format(sensor_name))
  File "/usr/local/lib/python2.7/logging/__init__.py", line 1212, in critical
    self._log(CRITICAL, msg, args, **kwargs)
  File "/usr/local/lib/python2.7/logging/__init__.py", line 1286, in _log
    self.handle(record)
  File "/usr/local/lib/python2.7/logging/__init__.py", line 1296, in handle
    self.callHandlers(record)
  File "/usr/local/lib/python2.7/logging/__init__.py", line 1336, in callHandlers
    hdlr.handle(record)
  File "/usr/local/lib/python2.7/logging/__init__.py", line 759, in handle
    self.emit(record)
  File "/code/dragonfly/dragonfly/status_log_handlers/amqp_handler.py", line 61, in emit
    self.connection_to_alert.send_status_message(severity=severity,alert=record.msg)
AttributeError: 'Service' object has no attribute 'send_status_message'
```
2. A logger.info is issued - **no error**:
```
2016-11-16T21:01:30[INFO    ] dripline.core.service(316) -> received a message
2016-11-16T21:01:30[DEBUG   ] dripline.core.service(363) -> Acknowledging message 8
2016-11-16T21:01:30[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T21:01:30Z", "msgtype": 4, "payload": {"value_cal": 5.5, "value_raw": "2.75"}, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "53e8870c6496", "version": "v0.0.0", "commit": ""}}
2016-11-16T21:01:30[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T21:01:30Z', u'msgtype': 4, u'payload': {u'value_cal': 5.5, u'value_raw': u'2.75'}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'53e8870c6496', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T21:01:30[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T21:01:30[DEBUG   ] dripline.core.gogol(48) -> in process_message callback
2016-11-16T21:01:30[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T21:01:30Z", "msgtype": 4, "payload": {"value_cal": 5.5, "value_raw": "2.75"}, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "53e8870c6496", "version": "v0.0.0", "commit": ""}}
2016-11-16T21:01:30[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T21:01:30Z', u'msgtype': 4, u'payload': {u'value_cal': 5.5, u'value_raw': u'2.75'}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'53e8870c6496', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T21:01:30[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T21:01:30[INFO    ] dragonfly.implementations.sensor_logger(58) -> value not logged for <misnamed_endpoint>
2016-11-16T21:01:30[INFO    ] dripline.core.service(329) -> Ready for next message
```

**Comment by guiguem** [comment]
which version of dripline are you using?

**Comment by les67** [comment]
Dripline tag wp2.0.0_go1.2.2

**Comment by guiguem** [comment]
You should use the latest one wp2.0.0_python3.3.3, because this method was implemented there

**Comment by laroque** [comment]
beat me to it, dragonfly should, as a rule, be run against the latest tagged version of dripline (unless doing development in dripline).

**Comment by les67** [comment]
The latest version of dripline did solve that problem. Now, when the sensor logger service receives a message with the value of this `misnamed_endpoint` it broadcasts the logger.critical message and moves onto the next succesfully:
```
2016-11-16T21:45:12[INFO    ] dripline.core.service(316) -> received a message
2016-11-16T21:45:12[DEBUG   ] dripline.core.service(363) -> Acknowledging message 5
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T21:45:12Z", "msgtype": 4, "payload": {"value_cal": 5.5, "value_raw": "2.75"}, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "a1e070376bbd", "version": "v0.0.0", "commit": ""}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T21:45:12Z', u'msgtype': 4, u'payload': {u'value_cal': 5.5, u'value_raw': u'2.75'}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'a1e070376bbd', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T21:45:12[DEBUG   ] dripline.core.gogol(48) -> in process_message callback
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T21:45:12Z", "msgtype": 4, "payload": {"value_cal": 5.5, "value_raw": "2.75"}, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "a1e070376bbd", "version": "v0.0.0", "commit": ""}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T21:45:12Z', u'msgtype': 4, u'payload': {u'value_cal': 5.5, u'value_raw': u'2.75'}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'a1e070376bbd', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T21:45:12[DEBUG   ] dragonfly.implementations.sensor_logger(45) -> basic_deliver.routing_key is: sensor_value.misnamed_endpoint
2016-11-16T21:45:12[DEBUG   ] dragonfly.implementations.sensor_logger(48) -> re_out.groupdict() is: {'from': 'misnamed_endpoint'}
2016-11-16T21:45:12[CRITICAL] dragonfly.implementations.sensor_logger(61) -> endpoint with name "misnamed_endpoint" was not found in database hence failed to log its value; might need it to add to the db
2016-11-16T21:45:12[INFO    ] dripline.core.service(329) -> Ready for next message
```
It then receives a new message with the value of another, valid, endpoint and logs that value. Right after that it receives a message whose payload is the critical message that was sent out previously. When it tries to parse this for the sensor_name it obviously fails and the service crashes:
```
2016-11-16T21:45:12[INFO    ] dripline.core.service(316) -> received a message
2016-11-16T21:45:12[DEBUG   ] dripline.core.service(363) -> Acknowledging message 7
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T21:45:12Z", "msgtype": 4, "payload": "endpoint with name \"misnamed_endpoint\" was not found in database hence failed to log its value; might need it to add to the db", "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "a1e070376bbd", "version": "v0.0.0", "commit": ""}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T21:45:12Z', u'msgtype': 4, u'payload': u'endpoint with name "misnamed_endpoint" was not found in database hence failed to log its value; might need it to add to the db', u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'a1e070376bbd', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T21:45:12[DEBUG   ] dripline.core.gogol(48) -> in process_message callback
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T21:45:12Z", "msgtype": 4, "payload": "endpoint with name \"misnamed_endpoint\" was not found in database hence failed to log its value; might need it to add to the db", "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "a1e070376bbd", "version": "v0.0.0", "commit": ""}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T21:45:12Z', u'msgtype': 4, u'payload': u'endpoint with name "misnamed_endpoint" was not found in database hence failed to log its value; might need it to add to the db', u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'a1e070376bbd', u'version': u'v0.0.0', u'commit': u''}}
2016-11-16T21:45:12[DEBUG   ] dripline.core.message(148) -> msgtype is 4
2016-11-16T21:45:12[DEBUG   ] dragonfly.implementations.sensor_logger(45) -> basic_deliver.routing_key is: status_message.p8_alerts.slow_control_table
2016-11-16T21:45:12[DEBUG   ] dragonfly.implementations.sensor_logger(47) -> re_out is: None
2016-11-16T21:45:12[WARNING ] dripline.core.gogol(55) -> got an exception:
'NoneType' object has no attribute 'groupdict'
2016-11-16T21:45:12[CRITICAL] dripline.core.service(435) -> Service <slow_control_table> crashing with error message:
'NoneType' object has no attribute 'groupdict'
2016-11-16T21:45:12[ERROR   ] dripline.core.service(436) -> Traceback (most recent call last):
```
I don't exactly understand how binding an exchange to a service works, but my question is: Why are we even catching/receiving such a message with the sensor logger service? Shouldn't we only be looking out for messages with `basic_deliver.routing_key` of format `sensor_value.<some_sensor_name>`? The message that crashes the service has a `basic_deliver.routing_key` of `status_message.p8_alerts.slow_control_table`.

**Comment by laroque** [comment]
This service itself is only interested in sensor_value.#, but it also needs a binding to broadcast.# (so that it can receive broadcast commands). If that binding is being made against # instead of broadcast.#, that could explain the problem.

**Comment by guiguem** [comment]
```python
        # listen to status_message alerts channel
        kwargs.update({'exchange':'alerts','keys':['status_message.#.#']})
        Gogol.__init__(self, **kwargs)```
Ben, correct me if I am wrong, but this is how to make a Gogol listen to messages with specific keys in alerts: replace status_message.#.# with sensor_name.# should make SensorLogger only listen to the things you are interested in...

**Comment by guiguem** [comment]
The default is indeed '#', unless you update the kwargs this way...

**Comment by laroque** [comment]
I don't remember exactly how all of this was changed when the gogol restructure is done, but indeed this Gogol should be adding a binding for `sensor_name.#`

Since there seems to be some question about it, once that is working and it isn't seeing the `status_message.#` messages, we should make sure that it is still responding to broadcast commands (`dragonfly cmd broadcast.ping` for example)

**Comment by guiguem** [comment]
It is service which adds the binding to broadcast. 
You should then see it when initializating a service:
```
2016-11-14T14:00:30^[[37m[DEBUG   ] dripline.core.service(250) -> ^[[35mBinding requests to sweeper_multido with broadcast.#^[[0m
2016-11-14T14:00:30^[[37m[DEBUG   ] dripline.core.service(250) -> ^[[35mBinding requests to sweeper_multido with sweeper_multido.#^[[0m
2016-11-14T14:00:30^[[37m[DEBUG   ] dripline.core.service(250) -> ^[[35mBinding requests to sweeper_multido with sweeper_settings.#^[[0m
2016-11-14T14:00:30^[[37m[DEBUG   ] dripline.core.service(250) -> ^[[35mBinding requests to sweeper_multido with broadcast.#^[[0m
2016-11-14T14:00:30^[[37m[DEBUG   ] dripline.core.service(250) -> ^[[35mBinding requests to sweeper_multido with sweeper_multido.#^[[0m
2016-11-14T14:00:30^[[37m[DEBUG   ] dripline.core.service(250) -> ^[[35mBinding requests to sweeper_multido with sweeper_settings.#^[[0m
```

**Comment by les67** [comment]
Ok thanks for the notes. Including a binding to `sensor_value.#` in alerts solved the issue above. I then checked the binding to broadcast.# via `dragonfly cmd broadcast.ping -b rabbit_broker` but got an error: 
```
2016-11-16T23:43:13[INFO    ] dripline.core.service(316) -> received a message
2016-11-16T23:43:13[DEBUG   ] dripline.core.service(363) -> Acknowledging message 108
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T23:43:13Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "59ae1c9cb088", "version": "v0.0.0", "commit": ""}, "payload": {"values": []}}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T23:43:13Z', u'payload': {u'values': []}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'59ae1c9cb088', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-16T23:43:13[INFO    ] dripline.core.spimescape(63) -> request received by slow_control_table
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(148) -> handling request:{"timestamp": "2016-11-16T23:43:13Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "59ae1c9cb088", "version": "v0.0.0", "commit": ""}, "payload": {"values": []}}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(154) -> routing key specifier is: ping
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T23:43:13Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "59ae1c9cb088", "version": "v0.0.0", "commit": ""}, "payload": {"values": []}}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T23:43:13Z', u'payload': {u'values': []}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'59ae1c9cb088', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-16T23:43:13[INFO    ] dripline.core.endpoint(159) -> got a 9 request: {u'values': []}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(176) -> method is: <bound method SensorLogger.on_cmd of <dragonfly.implementations.sensor_logger.SensorLogger object at 0x7f6b55bc45d0>>
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(125) -> operation allowed because not locked
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(179) -> args are:
[]
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(180) -> kwargs are:
{'routing_key_specifier': 'ping'}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(258) -> args are: ()
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(259) -> kwargs are: {'routing_key_specifier': 'ping'}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(184) -> 
 endpoint method returned 

2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(197) -> request method execution complete
2016-11-16T23:43:13[CRITICAL] dripline.core.service(435) -> Service <slow_control_table> crashing with error message:
'NoneType' object has no attribute 'send_reply'
2016-11-16T23:43:13[ERROR   ] dripline.core.service(436) -> Traceback (most recent call last):
  File "/code/dripline/python/dripline/core/service.py", line 433, in run
    self._connection.ioloop.start()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 138, in start
    self.poller.start()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 433, in start
    self.poll()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 494, in poll
    self._handler(fileno, event, write_only=write_only)
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 322, in _handle_events
    self._handle_read()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 351, in _handle_read
    self._on_data_available(data)
  File "/usr/local/lib/python2.7/site-packages/pika/connection.py", line 1285, in _on_data_available
    self._process_frame(frame_value)
  File "/usr/local/lib/python2.7/site-packages/pika/connection.py", line 1365, in _process_frame
    self._deliver_frame_to_channel(frame_value)
  File "/usr/local/lib/python2.7/site-packages/pika/connection.py", line 976, in _deliver_frame_to_channel
    return self._channels[value.channel_number]._handle_content_frame(value)
  File "/usr/local/lib/python2.7/site-packages/pika/channel.py", line 792, in _handle_content_frame
    self._on_deliver(*response)
  File "/usr/local/lib/python2.7/site-packages/pika/channel.py", line 886, in _on_deliver
    body)
  File "/code/dripline/python/dripline/core/service.py", line 326, in on_message
    msg_type_handlers[message.msgtype](unused_channel, basic_deliver, properties, body)
  File "/code/dripline/python/dripline/core/spimescape.py", line 68, in on_request_message
    self.handle_request(channel, method, header, body)
  File "/code/dripline/python/dripline/core/endpoint.py", line 199, in handle_request
    self.service.send_reply(properties, reply)
AttributeError: 'NoneType' object has no attribute 'send_reply'
```
The service is initialized with a binding to broadcast.# and sensor_values.# now
```
2016-11-16T23:35:19[DEBUG   ] dripline.core.service(250) -> Binding requests to slow_control_table with broadcast.#
2016-11-16T23:35:19[DEBUG   ] dripline.core.service(250) -> Binding alerts to slow_control_table with sensor_value.#
2016-11-16T23:35:19[DEBUG   ] dripline.core.service(250) -> Binding requests to slow_control_table with broadcast.#
2016-11-16T23:35:19[DEBUG   ] dripline.core.service(250) -> Binding alerts to slow_control_table with sensor_value.#
```
I'm not sure yet what's causing the latest break

**Comment by guiguem** [comment]
Are you sure this is the way one ping broadcast?

**Comment by les67** [comment]
All other services had no issues with the ping broadcast.
```
root@59ae1c9cb088:/# dragonfly cmd broadcast.ping -b rabbit_broker                                                                                                                                         
warning: slack is only ever critical, setting that
broadcast.ping(ret:0): [status_multido]-> {u'values': [None]}
broadcast.ping(ret:0): [slow_control_snapshot_service]-> {u'values': [None]}
broadcast.ping(ret:0): [run_table_service]-> {u'values': [None]}
broadcast.ping(ret:0): [generic_daq]-> {u'values': [None]}
broadcast.ping(ret:0): [my_store]-> {u'values': [None]}
```
 For example the mimic glenlivet service:
```
2016-11-16T23:43:13[INFO    ] dripline.core.service(316) -> received a message
2016-11-16T23:43:13[DEBUG   ] dripline.core.service(363) -> Acknowledging message 1
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T23:43:13Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "59ae1c9cb088", "version": "v0.0.0", "commit": ""}, "payload": {"values": []}}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T23:43:13Z', u'payload': {u'values': []}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'59ae1c9cb088', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-16T23:43:13[INFO    ] dripline.core.spimescape(63) -> request received by my_store
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(148) -> handling request:{"timestamp": "2016-11-16T23:43:13Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "59ae1c9cb088", "version": "v0.0.0", "commit": ""}, "payload": {"values": []}}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(154) -> routing key specifier is: ping
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-16T23:43:13Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "59ae1c9cb088", "version": "v0.0.0", "commit": ""}, "payload": {"values": []}}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-16T23:43:13Z', u'payload': {u'values': []}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'59ae1c9cb088', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-16T23:43:13[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-16T23:43:13[INFO    ] dripline.core.endpoint(159) -> got a 9 request: {u'values': []}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(176) -> method is: <bound method Spimescape.on_cmd of <dripline.core.utilities.Spimescape object at 0x7fb1507e45d0>>
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(125) -> operation allowed because not locked
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(179) -> args are:
[]
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(180) -> kwargs are:
{'routing_key_specifier': 'ping'}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(258) -> args are: ()
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(259) -> kwargs are: {'routing_key_specifier': 'ping'}
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(184) -> 
 endpoint method returned 

2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(197) -> request method execution complete
2016-11-16T23:43:13[INFO    ] dripline.core.service(573) -> sending a reply
2016-11-16T23:43:13[INFO    ] dripline.core.service(580) -> reply sent
2016-11-16T23:43:13[DEBUG   ] dripline.core.endpoint(200) -> reply sent
2016-11-16T23:43:13[INFO    ] dripline.core.spimescape(71) -> request processing complete
{}
2016-11-16T23:43:13[INFO    ] dripline.core.service(329) -> Ready for next message
```

**Comment by guiguem** [comment]
@laroque is it because SensorLogger service is not an endpoint of a Spimescape? My guess is that every Gogol based service (such as pid_loop), which does not have a Spimescape should failed too.

**Comment by les67** [comment]
@laroque @guiguem I've continued looking into the failed handling of the broadcast.ping request to the sensor logger service with no further progress. What concerns me is that this is an attribute error `self.service.send_reply(properties, reply)` -> `AttributeError: 'NoneType' object has no attribute 'send_reply'`. Besides this, the original issue is solved. Do one of you have any more thoughts on this? What are some broadcast commands that we would like the sensor logger service to respond to for instance?

**Comment by les67** [comment]
If the current issue _is_ inherit to all gogol based services, perhaps it would be more appropriate to open a new issue for this.

**Comment by laroque** [comment]
If it is a broader problem, then certainly a dedicated issue for it makes sense.

Broadcast commands are part of the dripline standard: https://github.com/project8/dripline/wiki/Use-Protocol#broadcast-requests. Unless there is a very good reason otherwise, support for those should come automatically through class inheritance. The one case of exceptions currently, is set_condition(<int>), which will need to be implemented by any service which needs to respond.

The attribute error should help you track this down. It appears that for this object self.service is not being set. Every Endpoint derived object in a config file should have the self.service attribute set to point to the one Service derived object in that config. That includes the service itself (which should just have a reference to itself) and any providers. Ideally this would be taken care of in the same logic that assigns the service reference for everything, but depending on how that is implemented the service may need to be a special case wherein `self.service = self` is done manually.

**Comment by wcpettus** [comment]
I don't understand how this could be an issue with Gogol's generally - 
Gogol inherits from Spimescape, and Spimescape in its __init__ method calls add_endpoint which binds self.service

So I'm confused how anything can have self.service is None

**Comment by les67** [comment]
Resolved with pull request #80.
